feat(ktoaster): component reskin [KHCP-9005] (#2015)
Kong/kongponents: docs/components/toaster.md

> Once `ToastManager` is added as a global property, you can access it's methods via `this.$toaster`

Reviewer: > Once `ToastManager` is added as a global property, you can access it's methods via `this.$toaster` if using the Vue Options API.

## Arguments

KToaster is the underlying component rendered by the `ToastManager` instance, so all component properties are passed down via `ToastManager.open()` methods' arguments. The accepted argument type is `string` or object that is instance of `Toast`.

```ts
interface Toast {
  key?: any
  title: string
  message?: string
  appearance?: ToasterAppearance
  timeoutMilliseconds?: number
}
```

### title

Notification title. When passing an object to `$toaster.open()` method, the `title` property is required. When passing a string, it will be rendered as title.

<div class="horizontal-container">
  <KButton @click="$toaster.open({ title: 'Long Title Gets Truncated With Ellipsis' })">Open Toaster</KButton>
  <KButton @click="$toaster.open('String Will Become A Title')" appearance="secondary">Open Toaster</KButton>
</div>

```html
<KButton @click="$toaster.open({ title: 'Long Title Gets Truncated With Ellipsis' })">Open Toaster</KButton>
<KButton @click="$toaster.open('String Will Become A Title')" appearance="secondary">Open Toaster</KButton>
```

### message

> Because a long title gets truncated, it is only good for short notifications. When a longer notification needs to be shown to the user, `message` property becomes helpful.

Reviewer: See my comment on truncation

>
  Open Toaster
</KButton>
```

### appearance

Depending on the nature of notification, you might want to use different appearances. KToaster supports these 5 appearances:
* `info` (default)
* `success`
* `danger`
* `warning`
* `system`

<div class="horizontal-container">
  <KButton @click="$toaster.open({ title: 'Info', appearance: 'info' })">
    <InfoIcon />
    Info
  </KButton>
  <KButton @click="$toaster.open({ title: 'Success', appearance: 'success' })">
    <CheckCircleIcon />
    Success
  </KButton>
  <KButton
    @click="$toaster.open({ title: 'Danger', appearance: 'danger' })"
    appearance="danger"
  >
    <ClearIcon />
    Danger
  </KButton>
  <KButton @click="$toaster.open({ title: 'Warning', appearance: 'warning' })">
    <WarningIcon />
    Warning
  </KButton>
  <KButton
    @click="$toaster.open({ title: 'System', appearance: 'system' })"
    appearance="secondary"
  >
    <KongIcon />
    System
  </KButton>
</div>

```ts
$toaster.open({ title: 'Warning', appearance: 'warning' })
```

### timeoutMilliseconds

> The default timeout, in milliseconds, is `5000` (5 seconds) however you can change this to by passing the `timeoutMilliseconds` property.

Author: I didn't, it was 5 seconds before - the docs were wrong

  {{timeLeft > 3 ? 'Open Toaster' : `Closing in ${timeLeft} seconds` }}
</KButton>

```vue
<template>
  <KButton @click="openNotification(toasterOptions)">Open Toaster</KButton>
</template>

<script setup lang="ts">
import type { Toast } from '@kong/kongponents'

const toasterOptions: Toast = {
  appearance: 'success',
  timeoutMilliseconds: 3000,
  title: 'Notification',
  message: 'This toaster has a 3 second timeout'
}

const openNotification = (options: Toast | string) => {
  $toaster.open(options)
}
</script>
```

## Toaster State

You can view the current state of active toasters by calling `this.$toaster.toasters`. Click the buttons below to watch the state change

<div class="horizontal-container">
  <KButton @click="openNotification({ timeoutMilliseconds: 10000, title: 'Info Notification', appearance: 'info' })">
    Open Toaster
  </KButton>
  <KButton
    @click="openNotification({ title: 'Danger Notification', appearance: 'danger' })"
    appearance="danger"
  >
    Open Toaster
  </KButton>
  <KButton
    @click="openNotification('Basic Notification')"
    appearance="secondary"
  >
    Open Toaster
  </KButton>
</div>

<pre>

>   {{ JSON.stringify(toasters || [], null, 2) }}

Reviewer: I like the idea of showing this data, but should the container have a max-height so the page doesn't grow/shrink?

Author: Made it fixed height and scrollable

## Arguments

KToaster is the underlying component rendered by the `ToastManager` instance, so all component properties are passed down via `ToastManager.open()` methods' arguments. The accepted argument type is `string` or object that is instance of `Toast`.

```ts
interface Toast {
  key?: any
  title: string
  message?: string
  appearance?: ToasterAppearance
  timeoutMilliseconds?: number
}
```

### title

> Notification title. When passing an object to `$toaster.open()` method, the `title` property is required. When passing a string, it will be rendered as title.

Reviewer: ```suggestion
Notification title. When passing an object to `$toaster.open()` method, the `title` property is required. When passing only a string, the string will be rendered as the title.
```

Reviewer: Can you also add a note here that titles are truncated?

<div class="horizontal-container">
  <KButton @click="$toaster.open({ title: 'Long Title Gets Truncated With Ellipsis' })">Open Toaster</KButton>
  <KButton @click="$toaster.open('String Will Become A Title')" appearance="secondary">Open Toaster</KButton>
</div>

```html
<KButton @click="$toaster.open({ title: 'Long Title Gets Truncated With Ellipsis' })">Open Toaster</KButton>
<KButton @click="$toaster.open('String Will Become A Title')" appearance="secondary">Open Toaster</KButton>
```

### message

> Because a long title gets truncated, it is only good for short notifications. When a longer notification needs to be shown to the user, `message` property becomes helpful.

Reviewer: ```suggestion
The message prop allows for displaying longer strings of text to the user. This prop is good for more detailed messages, or displaying IDs, etc.

Because a long title gets truncated, it is only good for short notifications. When a longer notification needs to be shown to the user, the `message` property is ideal.
```

<KButton @click="$toaster.open({ 
  title: 'Long Title Gets Truncated With Ellipsis',
  message: 'Detailed message. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.' })"
>
  Open Toaster
</KButton>

```html
<KButton @click="$toaster.open({ 
  title: 'Long Title Gets Truncated With Ellipsis',
  message: 'Detailed message. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.' })"
>
  Open Toaster
</KButton>
```

### appearance

Depending on the nature of notification, you might want to use different appearances. KToaster supports these 5 appearances:

> * `info` (default)

Reviewer: nit: add empty line
```suggestion

* `info` (default)
```

  <KButton @click="$toaster.open({ title: 'Warning', appearance: 'warning' })">
    <WarningIcon />
    Warning
  </KButton>
  <KButton
    @click="$toaster.open({ title: 'System', appearance: 'system' })"
    appearance="secondary"
  >
    <KongIcon />
    System
  </KButton>
</div>

```ts
$toaster.open({ title: 'Warning', appearance: 'warning' })
```

### timeoutMilliseconds

The default timeout, in milliseconds, is `5000` (5 seconds) however you can change this to by passing the `timeoutMilliseconds` property.

<KButton :disabled="timeLeft <= 3" @click="openNotificationElapse({ timeoutMilliseconds: 3000, appearance: 'success', title: 'Notification', message: `This toaster has a 3 second timeout` })">
  {{timeLeft > 3 ? 'Open Toaster' : `Closing in ${timeLeft} seconds` }}
</KButton>

```vue
<template>
  <KButton @click="openNotification(toasterOptions)">Open Toaster</KButton>
</template>

<script setup lang="ts">
import type { Toast } from '@kong/kongponents'

const toasterOptions: Toast = {
  appearance: 'success',
  timeoutMilliseconds: 3000,
  title: 'Notification',
  message: 'This toaster has a 3 second timeout'
}

const openNotification = (options: Toast | string) => {

>   $toaster.open(options)

Reviewer: I don't think this example works when using the Composition API? Same comment for all instances.

Reviewer: I think it would need to be something like this, showing how you can utilize via composable:

```typescript
import { ToastManager } from '@kong/kongponents'
import type { ToasterNotification } from '@kong/kongponents'

// Initialize the ToastManager as a singleton so a new instance isn't created each time
const toaster = new ToastManager()

export default function useToaster() {
  const notify = async (notification: Partial<ToasterNotification>): Promise<void> => {
    toaster.open({
      ...notification,
    })
  }

  return {
    notify,
  }
}
```

Author: It works and it is how we have it in script section of this file. This example refers to `timeoutMilliseconds` argument.

Reviewer: In this file yes because it's using the Options API right?

Author: No, it's using script setup with `getCurrentInstance` (screenshot). Ultimately, I think you are right and this component docs need to be updated to feature composable approach - I just don't really want to do it in this PR - how about we do it in `beta` when we refactor it not to create a second app instance?

![Screenshot 2024-02-16 at 11 07 07 AM](https://github.com/Kong/kongponents/assets/36751160/c5c4365f-c538-4a72-a6e0-d23763761a08)

Reviewer: works for me 👍🏼
